### PR TITLE
🐛  [Story preload] Fix preloading and specifying page in hashParam

### LINKED ISSUE #29351: Story branching URL param does not work from prerendered doc
Story branching URL param does not work from prerendered doc.

There are two bugs in one:

1. When prerendered, the first page is always the one that's shown
    - Side effect: If the page that's supposed to be loaded from a previous session is not the first page, when the document becomes visible it can break the page positioning and story state on desktop panels. 

2. When the document becomes visible, it doesn't play the right page

Repro URL: https://www.gstatic.com/amphtml/stamp/qa/branching.html#page=page-2&visibilityState=prerender

cc @ampproject/wg-stories

### PR DESCRIPTION
**Pull request by mszylkowski**
When you specify `#visibilityState=prerender&page=page-id` (example [here](https://www.gstatic.com/amphtml/stamp/qa/branching.html#page=page-2&visibilityState=prerender)) the second page is not loaded / shown, but the first one is. Fixes #29351

Demo: https://stories-demos-matias.web.app/examples/amp-story/interactives.html#visibilityState=prerender&page=quiz-intl

The fix involves 3 things:
- Set the hashParam'ed page as active instead of the first page by default when you pass a page in the hashString.
- Allow prerendering of the hashParam'ed **page** before the visibility changes. This marks the page as built so it updates the classes `i-amphtml-built` (for instance so that the text is not set to transparent) and it also triggers the `buildCallback` of the hashParam'ed page.
- Allow prerendering of the hashParam'ed page **layers** before the visibility changes. This is optional but sets the styles like applying the template classes (vertical, fill, etc) so the prerendered version doesn't look that bad.

Still, the previous page won't be in position, but it does show the correct active page.
image

### CODE REVIEW

**Comment by gmajoulet** [comment]
Can you please host a demo? There's a lot of things to check when it comes to resources loading, I can help you with that.

**Comment by mszylkowski** [comment]
> Can you please host a demo? There's a lot of things to check when it comes to resources loading, I can help you with that.

Added the demo at https://stories-demos-matias.web.app/examples/amp-story/interactives.html#visibilityState=prerender&page=quiz-intl

File: extensions/amp-story/1.0/amp-story.js [hunk 1]
```diff
@@ -410,8 +410,17 @@ export class AmpStory extends AMP.BaseElement {
...
+    let page = this.element.querySelector('amp-story-page');
+    const maybePageId = parseQueryString(this.win.location.hash)['page'];
+    const maybePage =
+      maybePageId &&
+      this.element.querySelector(
+        `amp-story-page#${escapeCssSelectorIdent(maybePageId)}`
+      );
+    if (maybePage) {
+      page = maybePage;
+    }
```
Comment (gmajoulet) [inline extensions/amp-story/1.0/amp-story.js:422]:
```suggestion
const page = this.element.querySelector(`amp-story-page#${escapeCssSelectorIdent(this.getInitialPageId_())}`)
```

File: extensions/amp-story/1.0/amp-story.js [hunk 2]
```diff
@@ -1036,9 +1045,9 @@ export class AmpStory extends AMP.BaseElement {
     // page is built. Other pages will only build if the document becomes
     // visible.
     if (!this.getAmpDoc().hasBeenVisible()) {
-      return whenUpgradedToCustomElement(firstPageEl).then(() => {
-        return firstPageEl.whenBuilt();
-      });
+      return whenUpgradedToCustomElement(firstPageEl).then(() =>
```
Comment (gmajoulet) [inline extensions/amp-story/1.0/amp-story.js:1048]: Not needed, please remove

File: extensions/amp-story/1.0/amp-story-page.js [hunk 3]
```diff
@@ -389,6 +394,21 @@ export class AmpStoryPage extends AMP.BaseElement {
...
+   * @return {boolean}
+   */
+  isPrerenderActivePage() {
+    if (this.isLoadedPage_ === null) {
+      const hashId = parseQueryString(this.win.location.href)['page'];
+      const identSelector = `amp-story-page#${escapeCssSelectorIdent(hashId)}`;
+      console.log(parseQueryString(this.win.location.href));
+      this.isLoadedPage_ = hashId && matches(this.element, identSelector);
+    }
+    console.log(this.isLoadedPage_);
```
Comment (gmajoulet) [inline extensions/amp-story/1.0/amp-story-page.js:408]: please remove

File: extensions/amp-story/1.0/amp-story-grid-layer.js [hunk 4]
```diff
@@ -115,7 +135,7 @@ export class AmpStoryGridLayer extends AmpStoryBaseLayer {
 
   /** @override */
   prerenderAllowed() {
-    return this.isFirstPage();
+    return this.isFirstPage() || this.isPrerenderActivePage();
```
Comment (gmajoulet) [inline extensions/amp-story/1.0/amp-story-grid-layer.js:138]: Please factorize these methods so they only run one query.

- if `location.href` has a `page` parameter, run the "am I the initial page" query
- if not, run the "isFirstPage" query

File: extensions/amp-story/1.0/amp-story-page.js [hunk 5]
```diff
@@ -823,7 +843,7 @@ export class AmpStoryPage extends AMP.BaseElement {
 
   /** @override */
   prerenderAllowed() {
-    return this.isFirstPage();
+    return this.isFirstPage() || this.isPrerenderActivePage();
```
Comment (gmajoulet) [inline extensions/amp-story/1.0/amp-story-page.js:846]: Re: one query

File: extensions/amp-story/1.0/amp-story.js (hunk 1 above)
Comment (mszylkowski) [reply extensions/amp-story/1.0/amp-story.js:422]: If the pageId passed does not exist, then we need to fallback on the first page as the active page. For that reason your snippet wouldn't work 100%, but I'll simplify it in a similar manner to make it fallback properly

File: extensions/amp-story/1.0/amp-story.js [hunk 6]
```diff
@@ -410,8 +410,15 @@ export class AmpStory extends AMP.BaseElement {
...
-    pageEl && pageEl.setAttribute('active', '');
+    const firstPage = this.element.querySelector(`amp-story-page`);
+    const activePage = firstPage
+      ? this.element.querySelector(
+          `amp-story-page#${escapeCssSelectorIdent(
+            this.getInitialPageId_(firstPage)
+          )}`
+        )
+      : null;
+    activePage && activePage.setAttribute('active', '');
```
Comment (gmajoulet) [inline extensions/amp-story/1.0/amp-story.js:421]: I think this all breaks the `layoutCallback` which won't resolve in a prerendering scenario.

Here, we assume that the first page is the only one that can be prerendered:

https://github.com/ampproject/amphtml/blob/19c934b4bc04c008121759b2cc61e6f7f44cb679/extensions/amp-story/1.0/amp-story.js#L1035-L1042

Instead, you should move the querySelector to the `getInitialPageId_()` method, and update the snippet I just linked to so it waits for the actual first page.

**Review by gmajoulet — COMMENTED** [review]
We should strive to make the overall codebase easier and easier with each of our PRs. If every code change we send makes the code a bit more complex to follow, or hides some extra method calls behind a if or a ternary conditions, it can very quickly become impossible to maintain.

File: extensions/amp-story/1.0/amp-story-grid-layer.js [hunk 7]
```diff
@@ -104,6 +109,23 @@ export class AmpStoryGridLayer extends AmpStoryBaseLayer {
...
+    if (this.isActivePage_ === null) {
+      const hashId = parseQueryString(this.win.location.href)['page'];
+      const identSelector = `amp-story-page#${escapeCssSelectorIdent(
+        hashId
+      )} amp-story-grid-layer`;
+      this.isActivePage_ = hashId
+        ? matches(this.element, identSelector)
+        : this.isFirstPage();
+    }
+    return this.isActivePage_;
```
Comment (gmajoulet) [inline extensions/amp-story/1.0/amp-story-grid-layer.js:126]: I don't know why my comment didn't send the first time, let me type it again.

You're making this code much harder to follow than it should be. I know it only runs one query but you have two, and the second one is called by the fallback of a ternary condition.
Instead of just adding an extra "if" to existing code, sometimes it's easy and better to re-write a small portion and make things easy.

Here, we could have one single method with one single query:

```
isPrerenderActivePage_() {
  // early return if method already executed

  let selector = 'amp-story-page:first-of-type';

  // Support prerender with branching .
  const hashId = ...;
  if (hashId) {
    selector += `, amp-story-page#${hashId}`;
  }

  return matches(this.element, selector);
}
```

It's less lines, less methods, less `if`, so much easier to test, maintain, and understand for a new developer.

File: extensions/amp-story/1.0/amp-story.js (hunk 6 above)
Comment (mszylkowski) [reply extensions/amp-story/1.0/amp-story.js:421]: I don't think it won't resolve because the first page also gets prerendered, but waiting for the active page could be a good idea. That's what you meant by "actual first page" right?

Comment (gmajoulet) [reply extensions/amp-story/1.0/amp-story.js]: Only one page needs to prerender. Otherwise we're wasting bandwidth/CPU. If the first page prerenders even though you have a `pageId` hash then your code needs to be fixed

File: examples/amp-story/interactives.html [hunk 8]
```diff
@@ -47,10 +47,11 @@ <h2>What is the best story ever?</h2>
         <amp-story-page id="quiz-intl">
           <amp-story-grid-layer template="fill">
             <amp-img
-                src="img/roadshow.jpg"
+                src="https://images.unsplash.com/photo-1543793804-ace495842b0e?ixid=MXwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHw%3D&ixlib=rb-1.2.1&auto=format&fit=crop&w=934&q=80"
```
Comment (gmajoulet) [inline examples/amp-story/interactives.html:50]: Changes to this file are unrelated to the PR, please remove

File: extensions/amp-story/1.0/amp-story-grid-layer.js [hunk 9]
```diff
@@ -84,24 +87,33 @@ export class AmpStoryGridLayer extends AmpStoryBaseLayer {
...
-    return this.isFirstPage_;
+    const hashId = parseQueryString(this.win.location.href)['page'];
+    let selector = 'amp-story-page:first-of-type';
+    if (hashId) {
+      selector += `, amp-story-page#${escapeCssSelectorIdent(hashId)}`;
+    }
+    const parentPage = closest(
+      this.element,
+      (el) => el.tagName.toLowerCase() === 'amp-story-page'
+    );
```
Comment (gmajoulet) [inline extensions/amp-story/1.0/amp-story-grid-layer.js:112]: It's a micro optimization but in the prerender case we can be extra cautious: given that `amp-story-grid-layer` HAS to be a direct descendant of `amp-story-page`, you can use `this.element.parentElement`

File: extensions/amp-story/1.0/amp-story-grid-layer.js [hunk 10]
```diff
@@ -84,24 +87,33 @@ export class AmpStoryGridLayer extends AmpStoryBaseLayer {
...
+    this.isActivePage_ =
+      toArray(this.win.document.querySelectorAll(selector)).pop() ===
+      parentPage;
```
Comment (gmajoulet) [inline extensions/amp-story/1.0/amp-story-grid-layer.js:115]: This loops through the array + mutates it and creates a new one, it's doing too much work for no good reason.
Please use `nodeList[nodeList.length - 1]` to access the last element

File: extensions/amp-story/1.0/amp-story.js (hunk 6 above)
Comment (mszylkowski) [reply extensions/amp-story/1.0/amp-story.js]: Ok that should be what I'm doing right now

File: extensions/amp-story/1.0/amp-story-page.js [hunk 11]
```diff
@@ -244,7 +245,7 @@ export class AmpStoryPage extends AMP.BaseElement {
     );
 
     /** @private {?boolean}  */
-    this.isFirstPage_ = null;
+    this.isActivePage_ = null;
```
Comment (gmajoulet) [inline extensions/amp-story/1.0/amp-story-page.js:248]:
```suggestion
    this.isPrerenderActivePage_ = null;
```

active page has a different meaning

**Review by gmajoulet — APPROVED** [review]

File: extensions/amp-story/1.0/amp-story-page.js (hunk 11 above)
Comment (mszylkowski) [reply extensions/amp-story/1.0/amp-story-page.js]: Changed this on both files 👍

**Comment by gmajoulet** [comment]
🚢
